src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,760 @@
+    status: Union[NTStatus, int] = NTStatus.SUCCESS,
+) -> None:
+    """
+    prepare and transmit a SMB header and payload
+    so actually a full packet but focus of function on header construction
+
+    @param command: command name or id, defaults to same as received packet
+    @type command: L{str} or L{int}
+
+    @param packet: the packet, C{data} contains after-header data
+    @type packet: L{_base.SMBPacket}
+
+    @param status: packet status, an NTSTATUS code
+    @type status: L{int} or L{NTStatus}
+    """
+    # FIXME credit and signatures not supported yet
+    if packet.hdr is None:
+        packet.hdr = HeaderSync()
+    packet.hdr.flags |= FLAG_SERVER
+    packet.hdr.flags &= ~FLAG_RELATED
+    if command is not None:
+        if isinstance(command, str):
+            cmds = [c[0] for c in COMMANDS]
+            icommand = cmds.index(command)
+        elif isinstance(command, int):
+            icommand = command
+        packet.hdr.command = icommand
+    packet.hdr.status = status
+    packet.hdr.credit_request = 1
+    packet.data = _base.pack(packet.hdr) + packet.data
+    packet.send()
+
+
+def smb_negotiate(packet: _base.SMBPacket, resp_type: type) -> None:
+    # capabilities is ignored as a 3.1 feature
+    # as are final field complex around "negotiate contexts"
+    dialects = struct.unpack_from(
+        "<%dH" % packet.body.dialect_count,
+        packet.data,
+        offset=_base.calcsize(HeaderSync) + packet.body.size,
+    )
+    signing_enabled = (packet.body.security_mode & NEGOTIATE_SIGNING_ENABLED) > 0
+    # by spec this should never be false
+    signing_required = (packet.body.security_mode & NEGOTIATE_SIGNING_REQUIRED) > 0
+    desc = ""
+    if signing_enabled:
+        desc += "ENABLED "
+    if signing_required:
+        desc += "REQUIRED"
+    log.debug(
+        """
+NEGOTIATE
+---------
+size            {sz}
+dialect count   {dc}
+signing         0x{sm:02x} {desc}
+client UUID     {uuid!r}
+dialects        {dlt!r}""",
+        sz=packet.body.size,
+        dc=packet.body.dialect_count,
+        sm=packet.body.security_mode,
+        desc=desc,
+        uuid=packet.body.client_uuid,
+        dlt=["%04x" % x for x in dialects],
+    )
+    negotiateResponse(packet, dialects)
+
+
+def errorResponse(packet: _base.SMBPacket, ntstatus: Union[NTStatus, int]) -> None:
+    """
+    send SMB error response
+
+    @type packet: L{_base.SMBPacket}
+    @type ntstatus: L{int} or L{NTStatus}
+    """
+    packet.data = ERROR_RESPONSE_MAGIC
+    sendHeader(packet, status=ntstatus)
+    # pre 3.1.1 no variation in structure
+
+
+def negotiateResponse(
+    packet: _base.SMBPacket, dialects: Optional[Sequence[int]] = None
+) -> None:
+    """
+    send negotiate response
+
+    @type packet: L{_base.SMBPacket}
+
+    @param dialects: dialects offered by client, if C{None}, 2.02 used
+    @type dialects: L{list} of L{int}
+    """
+    log.debug("negotiateResponse")
+    blob_manager = packet.ctx["blob_manager"]
```

Comment:
I really want mypy to have some visibility into *why* we expect `packet.ctx` to have a `blob_manager` key and why we expect it to be a `BlobManager`.